### DISCUSSION TITLE
Unable to run rabbitmqctl utility as a non root or non rabbitmq user.

### DISCUSSION THREAD

**Discussion by kaushiksrinivas** [rabbitmq/rabbitmq-server#3240] [Other]
We are unable to run the rabbitmqctl tool, without being a root user or being a rabbitmq user. In our k8s containers, we do not wish to run this utility as a root or as a rabbitmq user. When looked at the contents of this script, there are explicit checks if the user is a root user or rabbitmq user, if not the script just terminates.

What is the reason behind this hard coded checks and any possible safe work arounds for this ?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
@kaushiksrinivas can you please provide some details on how RabbitMQ was provisioned, what are you trying to do and why?
This issue is very light on specifics and we do not guess in this community.

The only installation method that limits some commands to administrative users is the Debian and RPM packages. It is highly unusual to see any of those used on Kubernetes [in this day and age](https://github.com/rabbitmq/cluster-operator/).
CLI tools themselves never require any particular user.

**Reply by kaushiksrinivas** [reply]
Hi @michaelklishin 

Below are the details of the environment we are running rabbitmq on,
- Run as docker/pod on a kubernetes environment. (rabbitmq rpm is installed in the docker image)
- SecurityContext is being used and runAsUser is changed to any random user id.
- rabbitmq container is run as non root user.
- /etc/passwd file within the rabbitmq container does not have the user id chosen mapped to username "rabbitmq"

Now, once the rabbitmq container is up and running, if a user tries to run the rabbitmqctl utility within the container we get the below error and the script terminates
**Only root or rabbitmq can run rabbitmqctl**

Here at this point, the process of rabbitmq is running as a non root as well as non "rabbitmq" user. And since the rabbitmqctl is also executed from inside the pod, even the rabbitmqctl is not executed as root or "rabbitmq" user.

And we did check this script, where we find that the script does explicit checks if the (id -u matches to "0") or (id -un matches to "rabbitmq"). In our case both condition does not satisfy and hence the script terminates.

Our goal is to be able to run this script rabbitmqctl without being either a "root" user or "rabbitmq" user.

Let me know if need any other information to add.

**Comment by michaelklishin** [comment]
The check is performed by a wrapper script, `rabbitmq-script-wrapper`, used by [Debian](https://github.com/rabbitmq/rabbitmq-packaging/blob/v3.9.x/debs/Debian/debian/rules#L38) and [RPM packages](https://github.com/rabbitmq/rabbitmq-packaging/blob/v3.9.x/RPMS/Fedora/rabbitmq-server.spec#L91) only.

Besides general practice of limiting administrative tools to certain users (privileged and `rabbitmq` user), one of the very likely
reasons why this is enforced is to make sure that [CLI tools discover the shared secret](https://www.rabbitmq.com/cli.html#erlang-cookie).

Modern CLI tools support `--erlang-cookie` as the argument but it is not the most secure option to pass it on the command line
and it would be highly confusing to the user to require that flag in scenario where RabbitMQ's effective user and the
user that runs CLI tools differ (which can easily be the case with OS package managers).

We are not aware of any such limitations in environments that use the [RabbitMQ cluster Operator for Kubernetes](https://www.rabbitmq.com/kubernetes/operator/operator-overview.html).

**Reply by kaushiksrinivas** [reply]
Hi @michaelklishin 

Below are the details of the environment we are running rabbitmq on,
- Run as docker/pod on a kubernetes environment. (rabbitmq rpm is installed in the docker image)
- SecurityContext is being used and runAsUser is changed to any random user id.
- rabbitmq container is run as non root user.
- /etc/passwd file within the rabbitmq container does not have the user id chosen mapped to username "rabbitmq"

Now, once the rabbitmq container is up and running, if a user tries to run the rabbitmqctl utility within the container we get the below error and the script terminates
**Only root or rabbitmq can run rabbitmqctl**

Here at this point, the process of rabbitmq is running as a non root as well as non "rabbitmq" user. And since the rabbitmqctl is also executed from inside the pod, even the rabbitmqctl is not executed as root or "rabbitmq" user.

And we did check this script, where we find that the script does explicit checks if the (id -u matches to "0") or (id -un matches to "rabbitmq"). In our case both condition does not satisfy and hence the script terminates.

Our goal is to be able to run this script rabbitmqctl without being either a "root" user or "rabbitmq" user.

Let me know if need any other information to add.

**Comment by kaushiksrinivas** [comment]
Hi @michaelklishin 

There is always a chance for rabbitmq server to run as a user apart from "root" or "rabbitmq" right. In those case, enforcing these utilities to be run as one of those users still will not help right ? since the cookie of actual rabbitmq server can be stored in the home directory of the user rabbitmq is run with (not root / not rabbitmq). In those cases if the cookie is copied to the home directory of the user which is running the cli tools should be enough to solve these problems right ? instead of binding these tools to run as only root or rabbitmq user ?

Please correct if the understanding is incorrect.
